text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,443 @@
+
+## Alternatives
+
+### Alternative: Entity Slicing using Data Tries
+
+[RFC 74, Entity Slicing using Data Tries](https://github.com/cedar-policy/rfcs/pull/74) proposes the idea of an _entity manifest_ to ease the task of entity slicing. This manifest is a trie-based data structure that expresses which entity attributes and ancestors are needed for each possible action. This data structure is strictly more precise than the concept of level, meaning that it may require less entity data to be provided. For example, the first part of our Basic Example in this RFC (policies 1, 2, and 4 of TinyTodo), the determined entity manifest would indicate that for the `GetList` action, you would need to provide `resource.owner`, `resource.readers`, `resource.editors` and the ancestors of `principal`. This is less data than would be retrieved by the slicing algorithm proposed here: These are level-1 valid policies, so the generic algorithm would acquire the same data given in the manifest but also more, e.g., the ancestors of `resource` and the attributes of `principal`.
+
+Entity manifests are less effective as a _prescriptive_ mechanism that aims to limit the shape of allowable policies. For example, a Cedar policy and entity storage service might like to disallow uploading policies whose entity data is expensive to retrieve in the worst case. With levels, such a service can upper-bound this work by blocking schemas with a level greater than a threshold, say 3. Validation will then ensure that uploaded policies respect this upper bound. Manifests are unwieldy as a specification mechanism, since they are specialized to particular entity definitions and name particular types and attributes, of which there could be many. They'd also have to be updated as types evolved over time, while a level-based specification is less likely to change.
+
+We believe that ultimately level-based validation and entity manifests should coexist. Level-based validation is used to bound entity retrieval work and prevent pathological policies, while entity manifests are used to more efficiently define the needed entity data. We could imagine a deployment strategy in which we accept this RFC and perform generic entity slicing as described here, and then later implement entity manifests, and start performing slicing using those instead.
+
+### Alternative: Level as a validation parameter, not in the schema
+
+This RFC has suggested that the level should be specified in the schema. Alternatively, we could specify the level as a parameter to the validator itself, leaving schemas unchanged. The benefit of doing so is that entity slicing becomes an orthogonal concern: The schema specifies the expected type structure of data provided with a request, but not how much of that data is required. The drawback is that policy writers cannot look in one place to know the limits on the policies they can write.
```

Comment:
> The drawback is that policy writers cannot look in one place to know the limits on the policies they can write.

I don't think this is so awful. The error message from the validator will tell the author what the limit is, if they hit it. This also allows for more flexibility in granularity of level validation in the future without having to change the Cedar language.

Reply:
After implementing this (https://github.com/cedar-policy/cedar/pull/1146) I am inclined towards making this a parameter.

Arguments for:
- In the non-JSON schema format, `level` can occur at the top of any namespace. This makes finding `level` annoying. We error if `level` occurs more than once, but we could just take it as a single parameter to `validate`.
- The JSON schema format has the same issue, but JSON isn't ordered, so it feels more native.
- Schema format remains unchanged and we can add `level` to the schema later if desired. In particular, adding the per-entity `level` as described in alternatives would require new syntax (something like `@level(1)`, as proposed in the alternative). Delaying this decision will let us hear if more people would prefer that and a `global level` or something similar.

Arguments against:
- Keep data in one place.
- Easier to change schema than to change code (but the entity slicing code would still need to change).